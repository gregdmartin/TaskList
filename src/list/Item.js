import React from 'react';

/**
 * Represents an individual item. It currently always has a checkbox.
 * TODO: decompose into CheckedItem and NormalItem
 */
class Item extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      id: this.props.id,
      shortDesc: this.props.shortDesc,
      checked: this.props.checked,
      editing: false
    }
    this.descInput = React.createRef();
    this.handleClick = this.handleClick.bind(this);
    this.toggleEdit = this.toggleEdit.bind(this);
  }

  handleClick() {
    this.setState(state => ({
      checked: !this.state.checked
    }));
  }

  toggleEdit() {
    let newState = {};
    if(this.state.editing) {
      newState['shortDesc'] = this.descInput.current.value;
    }
    newState['editing'] = !this.state.editing;
    this.setState(newState);
  }

  render() {
    let display;
    if(this.state.editing) {
      display = <input ref={this.descInput} type="text" defaultValue={this.state.shortDesc} />
    } else {
      display = <span className="short-desc">{this.state.shortDesc}</span>
    }
    return (
      <li className="task-item" id={this.state.id}>
        <input type="checkbox" defaultChecked={this.state.checked} onClick={this.handleClick}/>
        {display}
        &nbsp;
        <span onClick={this.toggleEdit}>[...]</span>
      </li>
    )
  }
}

export default Item;