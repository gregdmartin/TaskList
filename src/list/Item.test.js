import React from 'react';
import {shallow, mount} from 'enzyme';

import Item from './Item';

describe('<Item />', () => {
  describe('constructor()', () => {
    it('correctly sets initial state', () => {
      const wrapper = shallow(<Item id="1-1" shortDesc="a desc" checked={true} />);

      const state = wrapper.instance().state;
      expect(state.shortDesc).toEqual('a desc');
      expect(state.checked).toBe(true);
      expect(state.editing).toBe(false);
    });

    it('defaults checked to false if not provided', () => {
      const wrapper = shallow(<Item id="1-1" shortDesc="a desc" />);

      expect(wrapper.instance().state.checked).toBe(false);
    });
  });

  describe('handleClick()', () => {
    it('changes checkbox state', () => {
      const wrapper = shallow(<Item id="1-1" shortDesc="a desc" checked={false} />);
      wrapper.instance().handleClick();

      expect(wrapper.instance().state.checked).toBe(true);
    });
  });

  describe('toggleEdit()', () => {
    it('can toggle between editing and regular state, and applies changes to description', () => {
      const wrapper = mount(<Item id="1-1" shortDesc="a desc" checked={false} />);
      expect(wrapper.instance().state.editing).toBe(false);

      wrapper.instance().toggleEdit();
      expect(wrapper.instance().state.editing).toBe(true);

      wrapper.instance().descInput.current.value = "new desc";
      wrapper.instance().toggleEdit();

      const state = wrapper.instance().state;
      expect(state.editing).toBe(false);
      expect(state.shortDesc).toEqual('new desc');
    });
  });

  describe('render()', () => {
    it('renders default state as expected', () => {
      const wrapper = shallow(<Item id="1-1" shortDesc="a desc" checked={true} />);

      expect(wrapper).toMatchSnapshot();
    });

    it('renders an input when editing', () => {
      const wrapper = mount(<Item id="1-1" shortDesc="a desc" checked={true} />);
      wrapper.instance().toggleEdit();
      wrapper.update();

      expect(wrapper).toMatchSnapshot();
    });
  });
});