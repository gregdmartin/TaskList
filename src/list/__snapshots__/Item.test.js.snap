// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> render() renders an input when editing 1`] = `
<Item
  checked={true}
  id="1-1"
  shortDesc="a desc"
>
  <li
    className="task-item"
    id="1-1"
  >
    <input
      defaultChecked={true}
      onClick={[Function]}
      type="checkbox"
    />
    <input
      defaultValue="a desc"
      type="text"
    />
     
    <span
      onClick={[Function]}
    >
      [...]
    </span>
  </li>
</Item>
`;

exports[`<Item /> render() renders default state as expected 1`] = `
<li
  className="task-item"
  id="1-1"
>
  <input
    defaultChecked={true}
    onClick={[Function]}
    type="checkbox"
  />
  <span
    className="short-desc"
  >
    a desc
  </span>
   
  <span
    onClick={[Function]}
  >
    [...]
  </span>
</li>
`;
